modules/google-cloud-storage/src/xtdb/google/cloud_storage.clj

```diff
@@ -1,14 +1,182 @@
+           java.io.File
+           java.nio.charset.StandardCharsets))
 
-(defn ->document-store {::sys/deps (::sys/deps (meta #'ds/->nio-document-store))
-                        ::sys/args (::sys/args (meta #'ds/->nio-document-store))}
-  [opts]
-  (ds/->nio-document-store opts))
+(defn get-blob [{:keys [^Storage storage-service bucket-name prefix]} blob-name]
+  (let [blob-id (BlobId/of bucket-name (str prefix blob-name))]
+    (some-> (.get storage-service blob-id)
+            (.getContent (into-array Blob$BlobSourceOption [])))))
 
-(defn ->checkpoint-store {::sys/deps (::sys/deps (meta #'cp/->filesystem-checkpoint-store))
-                          ::sys/args (::sys/args (meta #'cp/->filesystem-checkpoint-store))}
-  [opts]
-  (cp/->filesystem-checkpoint-store opts))
+(defn get-blob-to-file [{:keys [^Storage storage-service bucket-name prefix]} blob-name file-path]
+  (let [blob-id (BlobId/of bucket-name (str prefix blob-name))]
+    (some-> (.get storage-service blob-id)
+            (.downloadTo file-path (into-array Blob$BlobSourceOption [])))))
+
+(defn put-blob-bytes
+  ([{:keys [^Storage storage-service bucket-name prefix]} blob-name data]
+   (let [blob-id (BlobId/of bucket-name (str prefix blob-name))
+         blob-info (.build (BlobInfo/newBuilder blob-id))]
+     (.create storage-service blob-info data (into-array Storage$BlobTargetOption [])))))
+
+(defn put-blob-from-file
+  ([{:keys [^Storage storage-service bucket-name prefix]} blob-name file-path]
+   (let [blob-id (BlobId/of bucket-name (str prefix blob-name))
+         blob-info (.build (BlobInfo/newBuilder blob-id))]
+     (.createFrom storage-service blob-info file-path (into-array Storage$BlobWriteOption [])))))
+
+(defn list-blobs 
+  [{:keys [^Storage storage-service bucket-name prefix]} obj-prefix]
+   (let [list-prefix (str prefix obj-prefix)
+         list-blob-opts (into-array Storage$BlobListOption
+                                    (if (not-empty list-prefix)
+                                      [(Storage$BlobListOption/prefix (str prefix obj-prefix))]
+                                      []))]
+     (->> (.list storage-service bucket-name list-blob-opts)
+          (.iterateAll)
+          (mapv (fn [^Blob blob]
+                  (subs (.getName blob) (count prefix)))))))
+
+(defn delete-blob [{:keys [^Storage storage-service bucket-name prefix]} blob-name]
+  (let [blob-id (BlobId/of bucket-name (str prefix blob-name))]
+    (.delete storage-service blob-id)))
+(defrecord GoogleCloudStorageDocumentStore [^Storage storage-service bucket-name prefix]
+  db/DocumentStore
+  (submit-docs [this docs]
+    (->> (for [[id doc] docs]
+           (future
+             ;; Put blob - key set to id, blob is nippy compressed document byte array
+             (put-blob-bytes this id (nippy/freeze doc))))
+         vec
+         (run! deref)))
+
+  (fetch-docs [this docs]
+    (xio/with-nippy-thaw-all
+      (reduce
+       #(if-let [doc (get-blob this %2)]
```

Comment:
nit: for these anon-fns, when you've got multiple arguments and you're using them multiple times, I tend to use `fn` with named params rather than `#()`